ADAL/src/workplacejoin/mac/ADWorkPlaceJoinUtil.m

```diff
@@ -22,14 +22,292 @@
+        goto _error;
+    }
+    
+    if (!identity || !userPrincipalName || !certificateIssuer || !certificateSubject || !certificateData || !privateKey)
+    {
+        // The code above will catch missing security items, but not missing item attributes. These are caught here.
+        ADAuthenticationError* adError = [ADAuthenticationError unexpectedInternalError:@"Missing some piece of WPJ data" correlationId:correlationId];
+        if (error)
+            *error = adError;
+        goto _error;
+    }
+    
+    // We found all the required WPJ information.
+    info = [[ADRegistrationInformation alloc] initWithSecurityIdentity:identity
+                                                     userPrincipalName:userPrincipalName
+                                                     certificateIssuer:certificateIssuer
+                                                           certificate:certificate
+                                                    certificateSubject:certificateSubject
+                                                       certificateData:certificateData
+                                                            privateKey:privateKey];
+    SAFE_ARC_AUTORELEASE(info);
+    
+    // Fall through to clean up resources.
+    
+_error:
+    
+    SAFE_ARC_RELEASE(certificateSubject);
+    SAFE_ARC_RELEASE(certificateData);
+    
+    if (identity)
+        CFRelease(identity);
+    if (certificate)
+        CFRelease(certificate);
+    if (privateKey)
+        CFRelease(privateKey);
+    
+    return info;
+}
+
+
++ (void) copyCertificate:(NSUUID*)correlationId
+                identity:(SecIdentityRef __nullable * __nonnull)identity
+             certificate:(SecCertificateRef __nullable * __nonnull)clientCertificate
+       certificateIssuer:(NSString* __nullable * __nonnull)clientCertificateIssuer
+                   error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    OSStatus status = noErr;
+    ADAuthenticationError* adError = nil;
+    NSData* issuer = nil;
+    NSMutableDictionary *identityQuery = nil;
+    CFDictionaryRef result = NULL;
+    
+    *identity = nil;
+    *clientCertificate = nil;
+    if (error)
+        *error = nil;
+
+    *clientCertificate = [self copyWPJCertificateRef:correlationId error:&adError];
+    
+    if (adError)
+    {
+        if (error)
+            *error = adError;
+        
+        AD_LOG_ERROR_F(@"Failed to retrieve WPJ client certificate from keychain - ", adError.code, nil, @"%@ correlation id.", correlationId);
+        goto _error;
+    }
+    
+    // If there's no certificate in the keychain, adError won't be set since this isn't an error condition.
+    if (!*clientCertificate)
+        return;
+    
+    // In OS X the shared access group cannot be set, so the search needs to be more
+    // specific. The code below searches the identity by passing the WPJ cert as reference.
+    identityQuery = [[NSMutableDictionary alloc] init];
```

Comment:
I prefer using a statically defined dictionary for keychain queries, unless there are optional parameters that aren't always included. e.g.:

```
NSDictionary* query = @{
    kSecClass : kSecClassIdentity,
   ...
};
```

As an added benefit this will crash if you pass in nulls on required parameters, `setObject:forKey:` will not. This is preferred because unexpected behavior can occur when query parameters that were expected become wild cards instead.